tests/trainer/connectors/test_data_connector.py

```diff
@@ -66,3 +67,19 @@ def test_dataloader_source_request_from_module():
     module.foo.assert_not_called()
     assert isinstance(source.dataloader(), DataLoader)
     module.foo.assert_called_once()
+
+
+def test_eval_distributed_sampler_warning(tmpdir):
+    """Test that a warning is raised when `DistributedSampler` is used with evaluation."""
+
+    model = BoringModel()
+    trainer = Trainer(strategy="ddp", devices=2, accelerator="cpu", fast_dev_run=True)
+    trainer._data_connector.attach_data(model)
+
+    with pytest.warns(UserWarning, match="use single device strategy to ensure each sample"):
+        trainer.state.fn = TrainerFn.VALIDATING
+        trainer.reset_val_dataloader(model)
+
+    with pytest.warns(UserWarning, match="use single device strategy to ensure each sample"):
+        trainer.state.fn = TrainerFn.TESTING
+        trainer.reset_test_dataloader(model)
```

Comment:
nit: cover only the minimum necessary

```suggestion
    trainer.state.fn = TrainerFn.VALIDATING
    with pytest.warns(UserWarning, match="use single device strategy to ensure each sample"):
        trainer.reset_val_dataloader(model)

    trainer.state.fn = TrainerFn.TESTING
    with pytest.warns(UserWarning, match="use single device strategy to ensure each sample"):
        trainer.reset_test_dataloader(model)
```